vj4/db.py

```diff
@@ -1,29 +1,36 @@
 import aiomongo
 import functools
 import urllib.parse
+from yarl import URL
 
 from vj4.util import options
 
 options.define('db_host', default='localhost', help='Database hostname or IP address.')
+options.define('db_port', default=27017, help='Database port.')
 options.define('db_name', default='test', help='Database name.')
 options.define('db_username', default='', help='Database username.')
 options.define('db_password', default='', help='Database password.')
+options.define('db_auth_source', default='',
+               help='Database name associated with the user\'s credential.')
 
 
 async def init():
   global _client, _db
 
-  def escape(s):
-    return urllib.parse.quote(s, safe='')
-
-  if not options.db_username:
-    uri = 'mongodb://{}'.format(escape(options.db_host))
-  else:
-    uri = 'mongodb://{}:{}@{}/?authSource={}'.format(escape(options.db_username),
-                                                     escape(options.db_password),
-                                                     escape(options.db_host),
-                                                     escape(options.db_name))
-  _client = await aiomongo.create_client(uri)
+  url_parts = {
+    'scheme': 'mongodb',
+    'host': options.db_host,
+    'path': options.db_name,
+    'port': options.db_port,
+    'user': options.db_username,
+    'password': options.db_password,
+    'query': {}
+  }
+  if options.db_auth_source:
+    url_parts['query']['authSource'] = options.db_auth_source
+  url = URL.build(**url_parts)
+
+  _client = await aiomongo.create_client(str(url))
   _db = _client.get_database(options.db_name)
```

Comment:
you can use `get_default_database()` here since you specified a default database in URL.